### PR TITLE
Use FQDN regex LRU everywhere

### PR DESCRIPTION
**Pull request by christarazi**
- defaults: Define FQDN regex compile LRU size default
- fqdn/re: Provide new package for accessing compile regex objects
- fqdn: Use new regex LRU package everywhere
- daemon, fqdn: Provide method for non-cached validation

See commit msgs.

Related: https://github.com/cilium/cilium/pull/17224
Related: https://github.com/cilium/cilium/pull/19383

### CODE REVIEW

File: pkg/fqdn/matchpattern/matchpattern.go
Comment (christarazi) [inline pkg/fqdn/matchpattern/matchpattern.go]: Maybe worth calling it out here, but reviewers might wonder: how do we know if this will even use the LRU, i.e. if the LRU even initialized at the time this code path executes?

The answer is yes it will use the LRU considering that this code path is called to "sanitize" the policies (read validate) upon submission from the user.

The LRU is initialized during dameon bootstrap of the FQDN subsystem which happens before the K8s watcher sync starts (`InitK8sSubsystem()`), so before Cilium can even process the policies applied by the user.

File: pkg/fqdn/re/re.go
Comment (joestringer) [inline pkg/fqdn/re/re.go]: Did you consider returning an error here?

I suspect that we never intend for this path to really be used, it's more of a programming error that indicates `InitRegexCompileLRU()` was not called early enough. Returning an error could help to identify this case and ensure that we just move the initialization earlier.

File: pkg/fqdn/matchpattern/matchpattern.go
Comment (christarazi) [reply pkg/fqdn/matchpattern/matchpattern.go]: Happy to add this as a comment somewhere or maybe inside the commit msg, once CI finishes.

File: pkg/fqdn/re/re.go
Comment (joestringer) [inline pkg/fqdn/re/re.go]: I'm not seeing any form of locking here, how do we ensure this is safe between goroutines/threads? Does the underlying library provide any guarantees in this regard?

File: pkg/fqdn/dnsproxy/proxy.go
Comment (joestringer) [inline pkg/fqdn/dnsproxy/proxy.go]: Any particular reason this doesn't follow the typical go form?

```suggestion
		rei, err := re.CompileRegex(mp)
		if err {
			return nil, err
		}
		newRE[selector] = rei
```

**Review by joestringer — COMMENTED** [review]
I'm mainly unsure about the locking requirements particularly if this is to be used again from another goroutine now.

File: pkg/fqdn/matchpattern/matchpattern.go
Comment (joestringer) [reply pkg/fqdn/matchpattern/matchpattern.go]: Ah hehe, exactly the topic of my feedback below: https://github.com/cilium/cilium/pull/19632#discussion_r862192900

File: pkg/fqdn/re/re.go
Comment (christarazi) [reply pkg/fqdn/re/re.go]: :scream: I somehow missed this. Good catch, I don't know why I didn't check my assumption on that. I assumed that a library like this would include its own locking.

```go
// Cache is an LRU cache. It is not safe for concurrent access.
type Cache struct {
```

File: pkg/fqdn/dnsproxy/proxy.go
Comment (christarazi) [reply pkg/fqdn/dnsproxy/proxy.go]: No, probably a leftover from iterating on the design of `CompileRegex`. Will simplify.

File: pkg/fqdn/re/re.go
Comment (joestringer) [reply pkg/fqdn/re/re.go]: Also worth noting in particular, that even if we protect the read/write of the LRU, the underlying entries are also being returned. So there's also the question of whether the underlying regex object can be concurrently used by multiple goroutines. (Though I'm guessing maybe we already thought about this in the existing design...?)

Comment (joestringer) [reply pkg/fqdn/re/re.go]: To answer the last question:

https://pkg.go.dev/regexp#Regexp

> A Regexp is safe for concurrent use by multiple goroutines, except for configuration methods, such as Longest.

File: pkg/fqdn/re/re.go
Comment (christarazi) [reply pkg/fqdn/re/re.go]: This is the area that I kinda struggled to come to a conclusion on. What I opted for here at a high-level is to ignore that initialization didn't happen and simply pass-through straight into a regex compile, transparent to the caller. And that was the biggest motivation--to be transparent. But with that comes that cost of hiding when a human error was made.

I just realized while typing this out that we could have a sync.Once inside this case to initialize the cache if it wasn't, so that should catch any "premature" calls to `CompileRegex` that were done without initializing the LRU first (in case of human error or even otherwise).

I'll implement that, thanks.

Comment (christarazi) [reply pkg/fqdn/re/re.go]: Hm, the initialization function needs to take an int for the size which we don't necessarily have with `CompileRegex` and would be awkward to always pass the size of the LRU to it... :thinking:

**Comment by christarazi** [comment]
Noting that only the runtime failed for legitimate reasons and the rest of CI has passed. Pushing a change to fix the failure and addresses the comments from above.

<details> 
  <summary>Diff </summary>

```diff
diff --git a/pkg/fqdn/dnsproxy/proxy.go b/pkg/fqdn/dnsproxy/proxy.go
index 1281e9f433..2362c0799d 100644
--- a/pkg/fqdn/dnsproxy/proxy.go
+++ b/pkg/fqdn/dnsproxy/proxy.go
@@ -820,12 +820,10 @@ func GetSelectorRegexMap(l7 policy.L7DataMap) (CachedSelectorREEntry, error) {
 		}
 		mp := strings.Join(reStrings, "|")
 		rei, err := re.CompileRegex(mp)
-		switch {
-		case rei != nil && err == nil:
-			newRE[selector] = rei
-		case err != nil:
+		if err != nil {
 			return nil, err
 		}
+		newRE[selector] = rei
 	}
 
 	return newRE, nil
diff --git a/pkg/fqdn/re/re.go b/pkg/fqdn/re/re.go
index 2d2e6f1acb..f71cc8a1bc 100644
--- a/pkg/fqdn/re/re.go
+++ b/pkg/fqdn/re/re.go
@@ -8,19 +8,38 @@ package re
 import (
 	"fmt"
 	"regexp"
+	"sync"
 
 	lru "github.com/golang/groupcache/lru"
+
+	"github.com/cilium/cilium/pkg/lock"
+	"github.com/cilium/cilium/pkg/logging"
+	"github.com/cilium/cilium/pkg/logging/logfields"
+	"github.com/cilium/cilium/pkg/option"
 )
 
-// CompileRegex provides a conviennce wrapper around the LRU. It will either
+// CompileRegex provides a convenience wrapper around the LRU. It will either
 // return the cached regex object or compile the pattern into a regex object if
-// the LRU hasn't been initialized or it doesn't exist in the cache. If it
-// doesn't exist in the cache, it will be added.
+// the LRU hasn't been initialized (and initializes it) or it doesn't exist in
+// the cache. If it doesn't exist in the cache, it will be added.
 func CompileRegex(p string) (*regexp.Regexp, error) {
 	if RegexCompileLRU == nil {
+		// Try to initialize the LRU here in case the caller never called
+		// InitRegexCompileLRU() explicitly.
+		initOnce.Do(func() {
+			if err := InitRegexCompileLRU(option.Config.FQDNRegexCompileLRUSize); err != nil {
+				log.WithError(err).Warn(
+					"Failed to initialize FQDN regex LRU. Cilium will not cache FQDN regexes. Please report this if you see this message.",
+				)
+			}
+		})
+		// Since this is the case where the LRU is not initialized, we accept
+		// that we won't cache the first result here.
 		return regexp.Compile(p)
 	}
+	RegexCompileLRU.Lock()
 	r, ok := RegexCompileLRU.Get(p)
+	RegexCompileLRU.Unlock()
 	if ok {
 		return r.(*regexp.Regexp), nil
 	}
@@ -28,7 +47,9 @@ func CompileRegex(p string) (*regexp.Regexp, error) {
 	if err != nil {
 		return nil, fmt.Errorf("failed to compile regex: %w", err)
 	}
+	RegexCompileLRU.Lock()
 	RegexCompileLRU.Add(p, n)
+	RegexCompileLRU.Unlock()
 	return n, nil
 }
 
@@ -38,18 +59,27 @@ func InitRegexCompileLRU(size int) error {
 		return fmt.Errorf("failed to initialize FQDN regex LRU due to invalid size %d", size)
 	}
 	RegexCompileLRU = &regexCompileLRU{
+		Mutex: &lock.Mutex{},
 		Cache: lru.New(size),
 	}
 	return nil
 }
 
-// RegexCompileLRU is a singleton instance of the LRU.
-var RegexCompileLRU *regexCompileLRU
+var (
+	// RegexCompileLRU is a singleton instance of the LRU.
+	RegexCompileLRU *regexCompileLRU
+	initOnce        sync.Once
+
+	log = logging.DefaultLogger.WithField(logfields.LogSubsys, "fqdn/re")
+)
 
-// regexCompileLRU is a shared LRU cache for storing compile regex objects of
+// regexCompileLRU is a shared LRU cache for storing compiled regex objects of
 // FQDN names or patterns, used in CiliumNetworkPolicy or
 // ClusterwideCiliumNetworkPolicy. This is meant to be a singleton that's
 // shared throughout Cilium.
 type regexCompileLRU struct {
+	// The lru package doesn't provide any concurrency guarantees so we must
+	// provide our own locking.
+	*lock.Mutex
 	*lru.Cache
 }
```

</details>

File: pkg/fqdn/re/re.go
Comment (christarazi) [reply pkg/fqdn/re/re.go]: Fixed by using option.Config.FQDNRegexCompileLRUSize. If the init fails, we'll log a warning. I don't expect to hit this case going forward as the initialization happens during the DNS proxy bootstrap. Assuming the shuffling of subsystem inits don't occur, this should never happen. If it does happen, we'll have a warning log. LMK if you think this addresses the concern or if we should think about a deeper approach to the problem.

File: pkg/fqdn/re/re.go
Comment (schlosna) [inline pkg/fqdn/re/re.go:80]: Relatedly, should this use a read/write mutex so that happy path cache hits only use read lock and don't block other readers, while cache miss acquires write lock?

```suggestion
	*lock.RWMutex
```

File: pkg/fqdn/re/re.go
Comment (schlosna) [inline pkg/fqdn/re/re.go]: Should this take out just a read lock to avoid blocking on read only cache hit happy path?

```suggestion
	RegexCompileLRU.RLock()
	r, ok := RegexCompileLRU.Get(p)
	RegexCompileLRU.RUnlock()
```

File: pkg/fqdn/re/re.go
Comment (joestringer) [reply pkg/fqdn/re/re.go]: Minor nit here, this now initializes the LRU on the first pass, but if the initialization is successful it will actually not cache that one first compiled regex. That said, this is also not such a big deal since if the same regex gets used many times, then it will end up in the LRU soon anyhow.

File: pkg/fqdn/re/re.go
Comment (joestringer) [reply pkg/fqdn/re/re.go]: I think that the concern here is that since the underlying structure is an LRU, presumably the `Get()` operation also modifies the underlying order that is stored, so readlocks would not be safe for concurrent access.

Overall though I presume that this will be primarily used during policy calculation and then the caller stores the underlying regex object for regular usage against the target requests without holding this lock, so it shouldn't be too bad even with a regular mutex.

File: pkg/fqdn/re/re.go
Comment (joestringer) [inline pkg/fqdn/re/re.go]: I think that even the global check here should be protected by the mutex, to ensure that this pointer isn't modified by one goroutine while another goroutine is initializing it?

File: pkg/fqdn/re/re.go
Comment (joestringer) [reply pkg/fqdn/re/re.go]: Hmm. Multiple paths for initialization is a bit awkward for some of the other reasons I've just posted. I think that this ends up making the code more complicated in a sort of "defensive coding" kind of way. If we can just guarantee that the structure is initalized early enough, then we don't need to worry about any of this extra logic.

Comment (christarazi) [reply pkg/fqdn/re/re.go]: We can guarantee that the DNS proxy in a non-test context will always initialize the LRU before serving DNS responses because we've put the LRU init inside the DNS proxy init.

What we can't really guarantee is if the daemon init ordering is for some reason shuffled around such that the K8s resources start syncing before the DNS proxy is up (and LRU is init'd). In that case, the LRU will not be initialized at first so we could be compiling regexes without caching them for a period of time, until the DNS proxy init starts. 

I'm not sure there is any way around this, at least until we have refectored the daemon init such that we can explicitly control which steps must happen before others, and so on.

This is why the original code was checking if the LRU was nil and just passing straight down to the regex compile call without doing anything else, because we assume that eventually, as long as we aren't in the testing context, that the LRU will get init'd.

Hopefully that clarifies this issue. So if the defensive coding is not really worth it, happy to remove it and go back to the original code as mentioned above.

File: pkg/fqdn/re/re.go
Comment (christarazi) [reply pkg/fqdn/re/re.go]: ^ Correct. If you look at the `Get()` call, you can see it's potentially modifying the underlying structure to evict the oldest member, etc.

File: pkg/fqdn/re/re.go
Comment (christarazi) [reply pkg/fqdn/re/re.go]: Good point, I've opted to use `atomic.{Load,Store}Pointer()`, instead of creating a mutex that will only be used once.

File: pkg/fqdn/re/re.go
Comment (joestringer) [reply pkg/fqdn/re/re.go]: > What we can't really guarantee is if the daemon init ordering is for some reason shuffled around such that the K8s resources start syncing before the DNS proxy is up (and LRU is init'd). In that case, the LRU will not be initialized at first so we could be compiling regexes without caching them for a period of time, until the DNS proxy init starts.

I was fishing for is a solution to this. If we don't provide this guarantee, then we could be spending more time/effort/memory on these regex compilations. For that undetermined period on startup, the cache does not work at all. Worst case, it makes the cache completely ineffective on startup and then it's only useful over time when new policies are added. My hope with the "error" suggestion would be that if the initialization returns an error up the stack, then at some point this will either trigger an agent-wide `log.Fatal()` or some sort of prominent error message to be logged. Then we catch this as part of testing in CI, detect whenever the guarantee is no longer provided, and make sure we fix it before that code gets into a release. That said, I'm not attached to this solution.

In my last comment, I was thinking that ideally there's just one place where we guarantee that the LRU is initialized. If we're going to make this Compile function more complicated to handle this case, then is it still useful to explicitly initialize the LRU from the DNS proxy? Presumably the first time that the DNS proxy uses this code, it will also call into the same Compile function. Is the DNS proxy initialization redundant, then? Maybe we can just contain the "worry about LRU initalization" logic to this Compile function?

(The other alternative is just explicitly initialize it super early even directly from the daemon init routines or something, but I'm not overly sold on that approach either).

Comment (christarazi) [reply pkg/fqdn/re/re.go]: Ok fair enough. Apologies this took longer than it should have. Yes returning an error will just simplify things. 

> My hope with the "error" suggestion would be that if the initialization returns an error up the stack, then at some point this will either trigger an agent-wide log.Fatal() or some sort of prominent error message to be logged. Then we catch this as part of testing in CI, detect whenever the guarantee is no longer provided, and make sure we fix it before that code gets into a release. That said, I'm not attached to this solution.

This didn't sound great to me at first because it's kind of a "stop the world" error, but it's not a bad idea now that I think about it a bit more. We won't run into this often anyway, if it all.

Comment (christarazi) [reply pkg/fqdn/re/re.go]: Implemented.

**Comment by christarazi** [comment]
@joestringer I've implemented the init comments along with the initial locking concerns with atomics. See change diff [here](https://github.com/cilium/cilium/compare/d8110b204d0a553742aeaae2cb6ae2c1be99f0b1..0ab4cd4151008503b7905b1b48dfbca5fd7bc394).

File: pkg/fqdn/re/re.go
Comment (schlosna) [reply pkg/fqdn/re/re.go]: Ahh, good call

**Review by joestringer — APPROVED** [review]

File: pkg/fqdn/re/re.go
Comment (joestringer) [reply pkg/fqdn/re/re.go]: :+1: No need to apologise, it's good practice to have fully discussion like this to try to figure out the ideal implementation or challenge our assumptions. You also can't read my mind unless I put it into words first :)

**Comment by christarazi** [comment]
Travis failed which exposed one code path where the LRU was uninitialized (resulting from diff [here](https://github.com/cilium/cilium/compare/d8110b204d0a553742aeaae2cb6ae2c1be99f0b1..0ab4cd4151008503b7905b1b48dfbca5fd7bc394)), this is fixed in this [diff](https://github.com/cilium/cilium/compare/01bd9cfdbcc522d134dd01bf531e117f47eab52d..1f822a8871df9b312dc97d88459c94166cc33643).

**Comment by christarazi** [comment]
More Travis failures exposing code paths, [diff](https://github.com/cilium/cilium/compare/1737f2ae02b310692524cd95db43a4ff0ead22d6..5fcbd45d303df66163d4bede236e32355470f73b).

**Comment by christarazi** [comment]
Fixed the last of the Travis failures due to uninitialized LRU (again), [diff](https://github.com/cilium/cilium/compare/5fcbd45d303df66163d4bede236e32355470f73b..b671673154e8938075ca735f39f04fe958aae12e).

Triggering the rest of CI.

**Review by joestringer — APPROVED** [review]

**Comment by christarazi** [comment]
Noting that all of CI passed except for the runtime tests and the conformance-test-ipv6 before the push.

Runtime failed legitimately, see latest [diff](https://github.com/cilium/cilium/compare/b671673154e8938075ca735f39f04fe958aae12e..462f0c72c290f68790be40f4616c7114144a8d6a). 

Not sure why conformance-test-ipv6 is failing, but saw errors around ` bpf-map-sync-cilium_lxc`.

**Comment by christarazi** [comment]
Pushing to rebase to see if that resolves conformance-test-ipv6 .

**Comment by christarazi** [comment]
Pushing a revision which moves away from the singleton LRU approach as pointed out by @aanm offline. Now the DNS proxy and the `matchpattern` package have their own instances of the cache, because both of these code paths could have a different number of entries inside them.

It's worth mentioning that we don't always want to insert into the cache, especially for code paths that handle the backend for user-invoked commands such as `cilium fqdn list|cache` because the user can pass a pattern which calls `matchpattern.Validate()`. This could evict entries from the cache for no reason, so we bypass the cache.

**Review by aanm — APPROVED** [review]

File: pkg/fqdn/re/re.go
Comment (joestringer) [inline pkg/fqdn/re/re.go]: nit, but typically when Cilium code creates an object it will begin with `New...()`. We can touch this up before merge though.

```suggestion
// NewRegexCompileLRU initializes the singleton LRU cache.
func NewRegexCompileLRU(size int) (*RegexCompileLRU, error) {
```

**Review by joestringer — APPROVED** [review]

File: daemon/cmd/daemon_main.go
Comment (joestringer) [inline daemon/cmd/daemon_main.go]: Hmm wait, this description says that it configures the FQDN MatchName / MatchPattern regex LRU size, but it actually configures the DNS L7 rule MatchName / MatchPattern. The FQDN one is being hardcoded to the `defaults.FQDNRegexCompileLRUSize`.

**Comment by christarazi** [comment]
All of CI passed except for the upgrade / downgrade test which is known to be failing at this moment. https://github.com/cilium/cilium/pull/19710 will fix that.

Pushing to address Joe's last comments, which have no functional changes.

**Comment by christarazi** [comment]
Latest push goes back to the singleton LRU approach after discussion with Joe and Andre. Summary is that there's no benefit for different code paths to have different LRUs, especially when they are policy validation and policy calculation. A policy will be applied by the user which may contain regexes to compile, we compile them as part of the validation, insert into the LRU, and then when endpoints are regeneration (policy recalculation), we use the LRU to fetch the regex for the expression in the policy. 

Separate LRUs would not allow the policy validation to share the regexes with the policy recalculation (until after the first policy recalculation). Hence why we go back to the singleton approach because it doesn't make much sense for this gap to exist.

File: pkg/fqdn/re/re.go
Comment (twpayne) [inline pkg/fqdn/re/re.go:34]: Why is `atomic.LoadPointer` needed here?

File: pkg/fqdn/re/re.go
Comment (twpayne) [inline pkg/fqdn/re/re.go:36]: Shouldn't this be a `panic` to catch misuse of this function early and loudly?

File: pkg/fqdn/re/re.go
Comment (twpayne) [inline pkg/fqdn/re/re.go:38]: Why not use a `sync.RWMutex` here? I saw another person made the same suggestion, but the suggestion was marked as resolved without comment or code change.

File: pkg/fqdn/re/re.go
Comment (twpayne) [inline pkg/fqdn/re/re.go:58]: Should `size == 0` not disable the cache completely?

File: pkg/fqdn/re/re.go
Comment (joestringer) [reply pkg/fqdn/re/re.go:80]: Same answer as here: https://github.com/cilium/cilium/pull/19632#discussion_r862384915

File: pkg/fqdn/re/re.go
Comment (christarazi) [reply pkg/fqdn/re/re.go:34]: See https://github.com/cilium/cilium/pull/19632#discussion_r862194361.

File: pkg/fqdn/re/re.go
Comment (christarazi) [reply pkg/fqdn/re/re.go:38]: See https://github.com/cilium/cilium/pull/19632#discussion_r869385577.

(I think there was some confusion with the PR being opened for v1.9 backport immediately and people reviewing this vs that PR.)

File: pkg/fqdn/re/re.go
Comment (christarazi) [reply pkg/fqdn/re/re.go:58]: The default is non-zero, so it would take someone explicitly to configure this to 0 or a developer explicitly calling this function with 0. I was mostly just mirroring the behavior of the library, but adding a warning in case someone did this without thinking. Do you suggest this becomes an error instead?

File: pkg/fqdn/re/re.go
Comment (christarazi) [reply pkg/fqdn/re/re.go:36]: Using panic is not something that we have a clear pattern for doing in Cilium. Yes we do have lots of Panic lines, but I'd venture to say those were all likely added to help with unit testing.

In this case, I avoided panic because this could result in Cilium crashing in a user environment and grinds the usage of Cilium to a halt. In particular, this code is called triggered from the following locations:
* on policy add
* on endpoint regeneration (policy recalculation)
* from the built-in CLI (`cilium policy ...`)
  * which is similar to the first point, because they're both API

Rather, I opted for an error that the user can share with us instead and we can fix. It's basically just a less intrusive panic. 

I agree in principle that panic gives us faster feedback by being more drastic, but I'm not sure that it's appropriate in this case.

**Comment by christarazi** [comment]
Rebased to re-run CI.

File: pkg/fqdn/re/re.go
Comment (joestringer) [reply pkg/fqdn/re/re.go:36]: In general we have shied away from `panic()` throughout the codebase specifically for this kind of resiliency reason. It's easy to introduce these statements with the best intentions, but then for them to slip through testing and end up causing production incidents in live environments not because of the condition that occurred, but because the agent is panicking.

The main exception is where the user has explicitly configured conflicting options, there we panic early to give the user feedback that this combination of settings is unsupported. The idea there is that as soon as the user installs / enables the option, they should get immediate feedback about the conflict.

**Comment by christarazi** [comment]
CI passed. Pushing to fix minor conflict with https://github.com/cilium/cilium/pull/19592 merged.

**Pull request merged**
